qualtran/bloqs/chemistry/sparse/prepare.py

```diff
@@ -132,6 +226,205 @@ def junk_registers(self) -> Tuple[SelectionRegister, ...]:
+            np.abs(integrals), epsilon=2**-num_bits_state_prep / num_non_zero
+        )
+        theta = (1 - np.sign(integrals)) // 2
+        num_lt = num_spin_orb // 2 * (num_spin_orb // 2 + 1)
+        one_body = np.array([0] * num_lt + [1] * len(integrals[num_lt:]))
+        alt_pqrs = indicies[alt]
+        alt_theta = theta[alt]
+        alt_one_body = one_body[alt]
+        return PrepareSparse(
+            num_spin_orb,
+            num_non_zero,
+            num_bits_state_prep,
+            tuple(tuple([int(_) for _ in x]) for x in alt_pqrs.T),
+            tuple([int(_) for _ in alt_theta]),
+            tuple([int(_) for _ in alt_one_body]),
+            tuple(tuple([int(_) for _ in x]) for x in indicies.T),
+            tuple([int(_) for _ in theta]),
+            tuple([int(_) for _ in one_body]),
+            tuple(keep),
+            num_bits_rot_aa=num_bits_rot_aa,
+            qroam_block_size=qroam_block_size,
+        )
+
+    def build_composite_bloq(
+        self,
+        bb: 'BloqBuilder',
+        d: 'SoquetT',
+        p: 'SoquetT',
+        q: 'SoquetT',
+        r: 'SoquetT',
+        s: 'SoquetT',
+        sigma: 'SoquetT',
+        alpha: 'SoquetT',
+        beta: 'SoquetT',
+        rot_aa: 'SoquetT',
+        swap_pq: 'SoquetT',
+        swap_rs: 'SoquetT',
+        swap_pqrs: 'SoquetT',
+        flag_1b: 'SoquetT',
+        alt_pqrs: 'SoquetT',
+        theta: 'SoquetT',
+        keep: 'SoquetT',
+        less_than: 'SoquetT',
+        alt_flag_1b: 'SoquetT',
+    ) -> Dict[str, 'SoquetT']:
+        # 1. Prepare \sum_d |d\rangle
+        d = bb.add(PrepareUniformSuperposition(self.num_non_zero), target=d)
+        # 2. QROM the ind_d alt_d values
+        n_n = (self.num_spin_orb // 2 - 1).bit_length()
+        target_bitsizes = (
+            (n_n,) * 4 + (1,) * 2 + (n_n,) * 4 + (1,) * 2 + (self.num_bits_state_prep,)
+        )
+        if self.qroam_block_size is None:
+            block_size = 2 ** find_optimal_log_block_size(self.num_non_zero, sum(target_bitsizes))
+        else:
+            block_size = self.qroam_block_size
+        qrom = SelectSwapQROM(
+            self.ind_pqrs[0],
+            self.ind_pqrs[1],
+            self.ind_pqrs[2],
+            self.ind_pqrs[3],
+            self.theta,
+            self.one_body,
+            self.alt_pqrs[0],
+            self.alt_pqrs[1],
+            self.alt_pqrs[2],
+            self.alt_pqrs[3],
+            self.alt_theta,
+            self.alt_one_body,
+            self.keep,
+            target_bitsizes=target_bitsizes,
+            block_size=block_size,
+        )
+        (
+            d,
+            p,
+            q,
+            r,
+            s,
+            theta[0],
+            flag_1b,
+            alt_pqrs[0],
+            alt_pqrs[1],
+            alt_pqrs[2],
+            alt_pqrs[3],
+            theta[1],
+            alt_flag_1b,
+            keep,
+        ) = bb.add(
```

Comment:
style: you could use `bb.add_d` which returns the output soquets as a dictionary and you can save some vertical space

Reply:
I'm not sure, wouldn't I have to do a lot of dictionary unpacking further on, like x=soqs['target12'], ...?

Reply:
yeah maybe it wouldn't make anything better